kms/managed.go

```diff
@@ -0,0 +1,224 @@
+package kms
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+const (
+	// CollisionTolerance with 2^21 as a very defensive value. 2^32 is more commonly used.
+	CollisionTolerance = 2097151
+	// keySize is the key size in bytes
+	keySize = 128 / 8
+	// nonceSize is the size of the nonce. Do not change, without breaking version change.
+	// The nonceSize is a de facto standard.
+	nonceSize = 12
+	// cacheSize is set to 100 keys, which can be used to decrypt up to 200 million encrypted
+	// files or data encrypted within nearly 2 years.
+	cacheSize = 100
+)
+
+var (
+	// ErrKeyExpired means that the expiration time of a key has come and it shouldn't be used any more.
+	ErrKeyExpired = errors.New("key is out of date and shouldn't be used anymore for encryption")
+	// ErrNoCipher means that there is no upstream kms given and therefore the keys in use can't be protected.
+	ErrNoCipher = errors.New("no upstream encryption service was specified")
+
+	week = time.Hour * 24 * 7
+)
+
+// ManagedCipher is a set of keys. Only one key is used for encryption at one
+// time. New keys are created automatically, when hitting safety thresholds.
+type ManagedCipher struct {
+	counter uint32
+	expires time.Time
+
+	keys            *cache
+	currentLocalKEK []byte
+
+	remoteKMSID    []byte
+	upstreamCipher EncrypterDecrypter
+
+	m sync.Mutex
+}
+
+// EncrypterDecrypter is a default encryption / decryption interface with an ID
+// to support remote state.
+type EncrypterDecrypter interface {
+	Encrypt(plaintext []byte) (keyID, ciphertext []byte, err error)
+	Decrypt(keyID, ciphertext []byte) (plaintext []byte, err error)
+}
+
+// CurrentKeyID returns the currently assumed remote Key ID.
+func (m *ManagedCipher) CurrentKeyID() []byte {
+	return m.remoteKMSID
+}
+
+// NewManagedCipher returns a pointer to a ManagedCipher. It is initialized with
+// a cache, a reference to an upstream cipher (like a KMS service or HMS) and
+// does an initial encryption call to the upstream cipher.
+func NewManagedCipher(upstreamCipher EncrypterDecrypter) (*ManagedCipher, error) {
+	if upstreamCipher == nil {
+		klog.Infof("create managed cipher without upstream encryption failed")
+		return nil, ErrNoCipher
+	}
+
+	mk := ManagedCipher{
+		keys:           newCache(cacheSize),
+		upstreamCipher: upstreamCipher,
+	}
+
+	if err := mk.manageKey(); err != nil {
+		klog.Infof("create managed cipher: %w", err)
+		return nil, err
+	}
+
+	klog.Infof("new managed cipher is created")
+
+	return &mk, nil
+}
+
+func (m *ManagedCipher) manageKey() error {
+	// Expensive lock that stops the world until we have a new key set up.
```

Comment:
Do we want to halt the world and wait for the state to be changed with a call to an external KMS?

Another more complex solution would be:
- We could just return an error without key rotation within `ManagedCipher`.
- The service or something else could create two instances of `ManagedCipher`. 
    - Instance 1, would be normal.
    - Instance 2, would have in addition to its own key, the key of instance 1.
- If instance 1 returns an error, we swap for instance 2 and continue.
    - We create another instance 3, which would have in addition to its own key, the keys of instance 2. 
- In the background we then merge / sync the caches.

Other suggestions are welcome.